File: ai-message-translator/backend/internal/ai/ai.go
package ai

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/Ocean1029/webapp/ai-message-translator/backend/internal/model"
	"github.com/anthropics/anthropic-sdk-go"
	"github.com/anthropics/anthropic-sdk-go/option"
)

// AnalysisRequest contains the input for a conversation analysis.
type AnalysisRequest struct {
	ConversationText string // Raw conversation text to analyze.
	ToneMode         string // "counselor" or "bestfriend".
}

// AnalysisResponse holds the structured result from Claude's analysis.
type AnalysisResponse struct {
	SubtextTranslation []model.SubtextEntry    `json:"subtextTranslation"`
	InterestScore      int                     `json:"interestScore"`
	ReplySuggestions   []model.ReplySuggestion `json:"replySuggestions"`
	Summary            string                  `json:"summary"`
}

// MessageCreateFunc is the function signature for creating Claude messages.
// It is extracted as a type to allow mocking in tests.
type MessageCreateFunc func(ctx context.Context, params anthropic.MessageNewParams, opts ...option.RequestOption) (*anthropic.Message, error)

// Client wraps the Anthropic SDK to provide conversation analysis.
type Client struct {
	createMessage MessageCreateFunc
}

// New creates a new AI client using the provided Anthropic API key.
func New(apiKey string) *Client {
	sdk := anthropic.NewClient(option.WithAPIKey(apiKey))
	return &Client{
		createMessage: sdk.Messages.New,
	}
}

// selectPrompt returns the appropriate system prompt based on the tone mode.
func selectPrompt(toneMode string) string {
	if toneMode == "bestfriend" {
		return bestFriendSystemPrompt
	}
	return counselorSystemPrompt
}

// Analyze sends the conversation text to Claude for analysis and returns
// the parsed response. It selects the system prompt based on the ToneMode
// field of the request.
func (c *Client) Analyze(ctx context.Context, req AnalysisRequest) (*AnalysisResponse, error) {
	systemPrompt := selectPrompt(req.ToneMode)

	msg, err := c.createMessage(ctx, anthropic.MessageNewParams{
		Model:     anthropic.ModelClaudeSonnet4_5,
		MaxTokens: 2048,
		System: []anthropic.TextBlockParam{
			{Text: systemPrompt},
		},
		Messages: []anthropic.MessageParam{
			anthropic.NewUserMessage(
				anthropic.NewTextBlock(req.ConversationText),
			),
		},
	})
	if err != nil {
		return nil, fmt.Errorf("claude API call failed: %w", err)
	}

	if len(msg.Content) == 0 {
		return nil, fmt.Errorf("claude returned empty response")
	}

	// Extract text from the first content block.
	block := msg.Content[0]
	if block.Type != "text" {
		return nil, fmt.Errorf("unexpected content block type: %s", block.Type)
	}

	var resp AnalysisResponse
	if err := json.Unmarshal([]byte(block.Text), &resp); err != nil {
		return nil, fmt.Errorf("failed to parse Claude response as JSON: %w", err)
	}

	return &resp, nil
}

File: ai-message-translator/frontend/src/lib/ai.ts
import Anthropic from "@anthropic-ai/sdk";

// System prompt: professional relationship counselor tone.
const counselorSystemPrompt = `You are a professional relationship counselor analyzing LINE chat conversations.
Analyze the conversation and respond in Traditional Chinese (Taiwan).

For each message from the other person, provide:
1. The subtext (what they likely mean beneath the surface)
2. An overall interest score from 1 to 10 (1 = not interested at all, 10 = extremely interested)
3. Two to three reply suggestions with expected effects
4. A brief summary of the overall conversation dynamics

You MUST respond with ONLY valid JSON in the following format (no markdown, no extra text):
{
  "subtextTranslation": [
    {"original": "the original message", "subtext": "what they really mean"}
  ],
  "interestScore": 7,
  "replySuggestions": [
    {"text": "suggested reply", "expectedEffect": "what effect this reply would have"}
  ],
  "summary": "brief overall analysis"
}`;

// System prompt: brutally honest best friend tone.
const bestFriendSystemPrompt = `You are the user's brutally honest best friend analyzing LINE chat conversations.
Be funny, use casual Taiwanese slang (e.g., \u6b38\u3001\u311f\u3001der\u3001\u8d85\u3001hen), and don't sugarcoat anything. Roast the situation if needed.

For each message from the other person, provide:
1. The subtext (roast-style translation of what they REALLY mean)
2. An overall interest score from 1 to 10 (1 = not interested at all, 10 = extremely interested)
3. Two to three reply suggestions ranging from bold to safe, with expected effects
4. A brief sarcastic summary of the overall situation

You MUST respond with ONLY valid JSON in the following format (no markdown, no extra text):
{
  "subtextTranslation": [
    {"original": "the original message", "subtext": "roast-style translation"}
  ],
  "interestScore": 5,
  "replySuggestions": [
    {"text": "suggested reply", "expectedEffect": "expected outcome"}
  ],
  "summary": "sarcastic summary of the situation"
}`;

// Structured response from Claude's analysis.
interface AnalysisResult {
  subtextTranslation: { original: string; subtext: string }[];
  interestScore: number;
  replySuggestions: { text: string; expectedEffect: string }[];
  summary: string;
}

// Select the system prompt based on tone mode.
function selectPrompt(toneMode: string): string {
  if (toneMode === "bestfriend") {
    return bestFriendSystemPrompt;
  }
  return counselorSystemPrompt;
}

// Strip markdown code block fences from Claude's response text.
function stripMarkdownFences(text: string): string {
  let cleaned = text.trim();
  if (cleaned.startsWith("```json")) {
    cleaned = cleaned.slice(7);
  } else if (cleaned.startsWith("```")) {
    cleaned = cleaned.slice(3);
  }
  if (cleaned.endsWith("```")) {
    cleaned = cleaned.slice(0, -3);
  }
  return cleaned.trim();
}

/**
 * Analyze a conversation using Claude AI.
 * Returns the structured analysis result.
 */
export async function analyzeConversation(
  text: string,
  toneMode: string
): Promise<AnalysisResult> {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) {
    throw new Error("ANTHROPIC_API_KEY environment variable is not set");
  }

  const client = new Anthropic({ apiKey });
  const systemPrompt = selectPrompt(toneMode);

  const message = await client.messages.create({
    model: "claude-sonnet-4-5",
    max_tokens: 2048,
    system: systemPrompt,
    messages: [
      {
        role: "user",
        content: text,
      },
    ],
  });

  if (message.content.length === 0) {
    throw new Error("Claude returned empty response");
  }

  const block = message.content[0];
  if (block.type !== "text") {
    throw new Error(`Unexpected content block type: ${block.type}`);
  }

  const cleaned = stripMarkdownFences(block.text);

  const result: AnalysisResult = JSON.parse(cleaned);
  return result;
}
